Lab4 remove psps

JamesCallaghan opened the pull request:
Update Lab 4:
- Remove Katacoda references - lab now requires students to run locally or in their own VM
- Use Gatekeeper instead of PSPs (PSS mentioned in the lab instructions, but Gatekeeper was chosen to show custom policy as an instructive exercise)
- Add exercise to build image and push to 'approved' repo (ttl.sh), which is then permitted via Rego Policy
- Build Gatekeeper Constraint Templates using Konstraint
- Add OPA unit tests

JamesCallaghan committed:
feat: update lab4 to use gatekeeper rather than PSPs
- update Makefile to include Gatekeeper configuration
- update kind config to use Kubernetes v1.25
- remove PSP references and files
- remove references to Katacoda

JamesCallaghan committed:
feat: add policy for approved image repo

- update makefile to use konstraint to generate gatekeeper constraint templates from rego
- add rego unit tests
- add a sample golang app demonstrating a multi-stage build, and push image to ttl.sh
- add policy only allowing images from our mock 'approved repo'
- add bats-detik tests for allowed as well as denied cases

JamesCallaghan committed:
style: remove old bats-detik tests

JamesCallaghan committed:
docs: update README with instructions for Lab 4

olcuhu committed:
fixed linux file permissions on lab4 to make it work on local machines

JamesCallaghan committed:
docs: fix typos in README

d1gital-f committed:
Minor modifications:
1) indentation (replaced spaces with tabs)
2) parameterised Gatekeeper and Kind versions
3) Added install-kind Makefile step

sublimino commented on lab4/local/Makefile:
```diff
@@ -1,38 +1,113 @@
+
+.PHONY: install-kind
+install-kind:
+	curl -Lo ${HOME}/kind https://kind.sigs.k8s.io/dl/${KIND_VERSION}/kind-linux-amd64
```
Should this be `$${KIND_VERSION}` as it's a Makefile variable and not an env var?

sublimino commented on lab4/local/Makefile:
```diff
@@ -1,38 +1,113 @@
+privileged-test: bats-detik-build
+	chmod 777 ./config && chmod -R 777 ./source 
+	docker run -it \
+		-v ${PWD}/source:/home/testing/source \
```
Worth quoting variables like `"${PWD}"`

JamesCallaghan replied:
great shout - addressed now

sublimino commented on lab4/local/README.md:
````diff
@@ -4,36 +4,114 @@
+```bash
+make install-gatekeeper
+```
+Enforcing policies using gatekeeper amounts to creating Constraint Templates and Constraints. We will use [Konstraint](https://github.com/plexsystems/konstraint) in order to generate Constraint Templates and Constraints using the Rego source files as the source of truth (meaning that we do not have to update the Rego and the YAML independently). Observe the Rego that we will use to generate our p ...
````
`gatekeeper` proper noun for consistency?

JamesCallaghan replied:
sounds good - addressed now

JamesCallaghan committed:
fix: add quotes in Makefile to variables which may contain spaces
- add quotes to variables
- add instructions for installing kind
- small stylistic changes in README

olcuhu reviewed (approved):
LGTM :fireworks:

sublimino commented on lab4/local/source/control-tests/allowed-repo/allowed-repo.yaml:
```diff
@@ -0,0 +1,15 @@
+    ports:
+    - containerPort: 80
+    securityContext:
+      privileged: false
```
Is it possible to drop all capabilities here too?

sublimino commented on lab4/local/source/control-tests/allowed-repo/non-allowed-repo.yaml:
```diff
@@ -0,0 +1,14 @@
+    ports:
+    - containerPort: 80
+    securityContext:
+      privileged: false
```
possible to drop caps?

sublimino commented on lab4/local/golang-app/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+WORKDIR /
+COPY --from=builder /hello_world /hello_world
+EXPOSE 8080
+ENTRYPOINT ["/hello_world"]
```
Which user does this end up running as, if `root` can we change it?

JamesCallaghan replied:
now running as the 'nonroot' user

sublimino commented on lab4/local/source/control-tests/privileged/privileged-test.bats:
```diff
@@ -0,0 +1,26 @@
+}
+
+@test "verify that privileged containers cannot be run in cluster" {
+  run kubectl apply -f source/control-tests/privileged/privileged.yaml
```
nit: identation

sublimino commented on lab4/local/source/control-tests/privileged/privileged-test.bats:
```diff
@@ -0,0 +1,26 @@
+
+@test "verify that non-privileged containers can be run in cluster" {
+	run kubectl apply -f source/control-tests/privileged/non-privileged.yaml
+	run verify "there is 1 pod named 'nginx-allowed'"
```
I'd like to test this if I can find time -- does it verify the pod is running? How many times will it try? There's advanced syntax in case of race conditions here https://github.com/bats-core/bats-detik/blob/0d71702f9016a955fc8197d562bb1bb88ddf63a8/README.md?plain=1#L158

JamesCallaghan replied:
In this basic example at the moment, it doesn't check the status of the pod, just that there is an appropriately named pod deployed to the cluster. We could check status and add a condition on number of tries similar to the line you've highlighted.

sublimino reviewed (approved):
Minor nits, looks fantastic! Awesome work @JamesCallaghan and team 🙏

JamesCallaghan committed:
fix: drop all caps and run sample app as nonroot user

- drop all capabilities for test pods
- run example golang app as nonroot user
- indentation consistency changes
